File: 130-Surrounded-Regions.py
class Solution:
    def solve(self, board: List[List[str]]) -> None:
        queue=deque([])
        for i in range(len(board)):
            for j in range(len(board[i])):
                if (i==0 or j==0 or i==len(board)-1 or j==len(board[i])-1) and board[i][j]=='O':
                    queue.append((i,j))
        values=[(0,1),(1,0),(-1,0),(0,-1)]
        visited=set()
        while queue:
            row,col=queue.popleft()
            board[row][col]=1
           
            visited.add((row,col))
            for val in values:
                r=row+val[0]
                c=col+val[1]
                if r>=0 and r<len(board) and c>=0 and c<len(board[r]):
                    if board[r][c]=='O':
                        queue.append((r,c))
                        visited.add((r,c))
        
        for i in range(len(board)):
            for j in range(len(board[i])):
                if board[i][j]==1:
                    board[i][j]='O'
                elif board[i][j]=='O':
                    board[i][j]='X'


        
        

File: 130-Surrounded-Regions.cpp
class Solution {
public:
    void solve(vector<vector<char>>& board) {
        queue<pair<int,int>> q;
        vector<pair<int,int>> dir={{-1,0},{1,0},{0,-1},{0,1}};
        for(int i=0;i<board.size();i++){
            for(int j=0;j<board[0].size();j++){
                if((i==0 || j==0 || i==board.size()-1 || j==board[0].size()-1) && board[i][j]=='O'){
                    q.push({i,j});
                }
            }
        }
        while(!q.empty()){
            int row=q.front().first;
            int col=q.front().second;
            board[row][col]='1';
            q.pop();
            for(auto it:dir){
                int c_row=row+it.first;
                int c_col=col+it.second;
                if(c_row>=0 && c_col>=0 && c_row<board.size() && c_col<board[0].size() && board[c_row][c_col]=='O'){
                    q.push({c_row,c_col});
                }

            }

        }
        for(int i=0;i<board.size();i++){
            for(int j=0;j<board[0].size();j++){
               
                if(board[i][j]=='1'){
                    board[i][j]='O';
                }
                else if(board[i][j]=='O'){
                    board[i][j]='X';
                }
            }
        }
    }
};